contracts/ColonyNetworkAuction.sol

```diff
@@ -0,0 +1,203 @@
+    public 
+    {
+    colonyNetwork = msg.sender;
+    require(_clnyToken != 0x0 && _token != 0x0);
+    clnyToken = ERC20Extended(_clnyToken);
+    token = ERC20Extended(_token);
+    
+    require(_startPrice > 0);
+    startPrice = _startPrice;
+    require(_quantity > 0);
+    quantity = _quantity;
+    startBlock = block.number;
+    AuctionStarted(startBlock);
+  }
+
+  function bid(uint256 _amount) public
+  auctionNotFinalized
+  endBlockNotSet
+  {
+    require(_amount > 0);
+    uint _totalToEndAuction = totalToEndAuction();
+    uint remainingToEndAuction = sub(_totalToEndAuction, receivedTotal);
+
+    // Adjust the amount for final bid in case that takes us over the offered quantity at current price
+    // Also conditionally set the endBlock
+    uint amount;
+    if (remainingToEndAuction > _amount) {
+      amount = _amount;
+    } else {
+      amount = remainingToEndAuction;
+      endBlock = block.number;
+    }
+    
+    clnyToken.transferFrom(msg.sender, this, amount);
+    bids[msg.sender] = add(bids[msg.sender], amount);
+    receivedTotal = add(receivedTotal, amount);
+    
+    AuctionBid(msg.sender, amount, sub(remainingToEndAuction, amount));
+  }
+
+  // Finalize the auction and sets the final Token price
+  function finalize() public
+  auctionNotFinalized
+  endBlockSet
+  {
+    // Give the network all CLNY sent to the auction in bids
+    clnyToken.transfer(colonyNetwork, receivedTotal);
+    finalPrice = mul(receivedTotal, MULTIPLIER) / quantity;
+    finalized = true;
+    AuctionFinalized(finalPrice);
+  }
+
+  function claim() public 
+  auctionFinalized
+  returns (bool)
+  {
+    uint amount = bids[msg.sender];
+    uint tokens = mul(amount, MULTIPLIER) / finalPrice;
+
+    // Due to finalPrice floor rounding, the number of assigned tokens may be higher
+    // than expected. Therefore, the number of remaining unassigned auction tokens
+    // may be smaller than the number of tokens needed for the last claimTokens call
+    uint auctionTokenBalance = token.balanceOf(address(this));
+    if (tokens > auctionTokenBalance) {
+      tokens = auctionTokenBalance;
+    }
+
+    // Update the total amount of funds for which tokens have been claimed
+    tokensClaimed += amount;
+
+    // Set receiver bid to 0 before transferring the tokens
+    bids[msg.sender] = 0;
+    require(token.transfer(msg.sender, tokens));
+    AuctionClaim(msg.sender, tokens);
+
+    assert(token.balanceOf(msg.sender) >= tokens);
+    assert(bids[msg.sender] == 0);
+    return true;
+  }
+
+  function totalToEndAuction() public view returns (uint) {
+    return (quantity * price()) / MULTIPLIER;
+  }
+
+  // Get the current Token price in CLNY wei
+  // If the end block is set, i.e. auction is closed for bids, use the endBlock in calculation
+  function price() public view returns (uint) {
+    uint lastBlock = endBlock == 0 ? block.number : endBlock;
+    return startPrice * MULTIPLIER / (1 + lastBlock - startBlock);
```

Comment:
I think a linear price decay function is too slow, and it's definitely the case if we start with an absurdly high start price, as I'd prefer.

Reply:
This is slow? It falls off a cliff in 10 blocks
![screen shot 2018-02-28 at 17 29 31](https://user-images.githubusercontent.com/703848/36795949-f81fb976-1cac-11e8-8e9c-a57b9e3d9581.png)